mempool/cat/pool.go

```diff
@@ -0,0 +1,765 @@
+// to all other peers.
+func (txmp *TxPool) next() <-chan types.TxKey {
+	txmp.broadcastMtx.Lock()
+	defer txmp.broadcastMtx.Unlock()
+	if len(txmp.txsToBeBroadcast) != 0 {
+		ch := make(chan types.TxKey, 1)
+		for key := range txmp.txsToBeBroadcast {
+			delete(txmp.txsToBeBroadcast, key)
+			ch <- key
+			return ch
+		}
+	}
+	return txmp.broadcastCh
+}
+
+// markToBeBroadcast marks a transaction to be broadcasted to peers.
+// This should never block so we use a map to create an unbounded queue
+// of transactions that need to be gossiped.
+func (txmp *TxPool) markToBeBroadcast(key types.TxKey) {
+	if !txmp.config.Broadcast {
+		return
+	}
+
+	select {
+	case txmp.broadcastCh <- key:
+	default:
+		txmp.broadcastMtx.Lock()
+		defer txmp.broadcastMtx.Unlock()
+		txmp.txsToBeBroadcast[key] = struct{}{}
+	}
+}
+
+// TryAddNewTx attempts to add a tx that has not already been seen before. It first marks it as seen
+// to avoid races with the same tx. It then call `CheckTx` so that the application can validate it.
+// If it passes `CheckTx`, the new transaction is added to the mempool as long as it has
+// sufficient priority and space else if evicted it will return an error
+func (txmp *TxPool) TryAddNewTx(tx types.Tx, key types.TxKey, txInfo mempool.TxInfo) (*abci.ResponseCheckTx, error) {
+	// First check any of the caches to see if we can conclude early. We may have already seen and processed
+	// the transaction if:
+	// - We are connected to nodes running v0 or v1 which simply flood the network
+	// - If a client submits a transaction to multiple nodes (via RPC)
+	// - We send multiple requests and the first peer eventually responds after the second peer has already provided the tx
+	if txmp.IsRejectedTx(key) {
+		// The peer has sent us a transaction that we have previously marked as invalid. Since `CheckTx` can
+		// be non-deterministic, we don't punish the peer but instead just ignore the tx
+		return nil, ErrTxAlreadyRejected
+	}
+
+	if txmp.WasRecentlyEvicted(key) {
+		// the transaction was recently evicted. If true, we attempt to re-add it to the mempool
+		// skipping check tx.
+		return txmp.TryReinsertEvictedTx(key, tx, txInfo.SenderID)
+	}
+
+	if txmp.Has(key) {
+		txmp.metrics.AlreadySeenTxs.Add(1)
+		// The peer has sent us a transaction that we have already seen
+		return nil, ErrTxInMempool
+	}
+
+	// reserve the key
+	if !txmp.store.reserve(key) {
+		txmp.logger.Debug("mempool already attempting to verify and add transaction", "txKey", fmt.Sprintf("%X", key))
+		txmp.PeerHasTx(txInfo.SenderID, key)
+		return nil, ErrTxInMempool
+	}
+	defer txmp.store.release(key)
+
+	// If a precheck hook is defined, call it before invoking the application.
+	if err := txmp.preCheck(tx); err != nil {
+		txmp.metrics.FailedTxs.Add(1)
+		return nil, mempool.ErrPreCheck{Reason: err}
+	}
+
+	// Early exit if the proxy connection has an error.
+	if err := txmp.proxyAppConn.Error(); err != nil {
```

Comment:
should we check the connection before running the preCheck function? Or does the ordering of these two checks not matter much? If so, why?

Reply:
to me the error with a connection would trump the error from the preCheck which is why i would think it connection check would be first.